src/app/modules/verification/verification.service.ts

```diff
@@ -256,80 +249,98 @@ export const resetFieldForTransaction = (
+  fieldId,
+  recipient,
+  otp,
+  hashedOtp,
+}: SendOtpParams): ResultAsync<
+  IVerificationSchema,
+  | TransactionNotFoundError
+  | DatabaseError
+  | FieldNotFoundInTransactionError
+  | TransactionExpiredError
+  | WaitForOtpError
+  | MalformedParametersError
+  | SmsSendError
+  | InvalidNumberError
+  | MailSendError
+  | NonVerifiedFieldTypeError
+> => {
+  const logMeta = {
+    action: 'sendNewOtp',
+    transactionId,
+    fieldId,
   }
-  const { _id: transactionId, formId } = transaction
-  // TODO (#317): remove usage of non-null assertion
-  const waitForSeconds = waitToResendOtpSeconds(field.hashCreatedAt!)
-  if (waitForSeconds > 0) {
-    return throwError(
-      `Wait for ${waitForSeconds} seconds before requesting for a new otp`,
-      VfnErrors.WaitForOtp,
-    )
-  } else {
-    const hashCreatedAt = new Date()
-    const otp = generateOtp()
-    const hashedOtp = await bcrypt.hash(otp, SALT_ROUNDS)
-
-    const signedData = formsgSdk.verification.generateSignature!({
-      transactionId,
-      formId,
-      fieldId,
-      answer,
-    })
+  return getValidTransaction(transactionId).andThen((transaction) =>
+    getFieldFromTransaction(transaction, fieldId)
+      .asyncAndThen((field) => {
+        if (!isOtpWaitTimeElapsed(field.hashCreatedAt)) {
+          logger.warn({
+            message: 'OTP requested before waiting time elapsed',
+            meta: logMeta,
+          })
+          return errAsync(new WaitForOtpError())
+        }
 
-    await sendOtpForField(formId, field, answer, otp)
+        return sendOtpForField(transaction.formId, field, recipient, otp)
+      })
       .andThen(() => {
+        const signedData = formsgSdk.verification.generateSignature({
+          transactionId,
+          formId: transaction.formId,
+          fieldId,
+          answer: recipient,
+        })
         return ResultAsync.fromPromise(
-          VerificationModel.updateOne(
-            { _id: transactionId, 'fields._id': fieldId },
-            {
-              $set: {
-                'fields.$.hashCreatedAt': hashCreatedAt,
-                'fields.$.hashedOtp': hashedOtp,
-                'fields.$.signedData': signedData,
-                'fields.$.hashRetries': 0,
-              },
-            },
-          ).exec(),
+          VerificationModel.updateHashForField({
+            fieldId,
+            hashedOtp,
+            signedData,
+            transactionId,
+          }),
           (error) => {
             logger.error({
               message:
-                'Database error occurred whilst updating verification document',
-              meta: {
-                action: 'getNewOtp',
-                formId,
-                fieldId,
-                transactionId,
-              },
+                'Error while updating transaction data after sending OTP',
+              meta: logMeta,
               error,
             })
             return new DatabaseError(getMongoErrorMessage(error))
```

Comment:
nit: return `transformMongoError` instead?

Reply:
ok seeing as `DatabaseError` is used everywhere, will make the changes in #1455 to apply this consistently